esp/platform/esptraceloggingcomponent.hpp

```diff
@@ -74,19 +74,38 @@ interface ITraceLoggingComponent : extends IInterface
     virtual LogMsgDetail tracePriorityLimit(const LogMsgCategory& category) const = 0;
 
     /// Directs accepted content to implemtation-defined targets.
-    virtual void         traceOutput(const LogMsgCategory& category, const char* format, va_list& arguments) const = 0;
+    virtual void         traceOutput(const LogMsgCategory& category, const char* format, va_list& arguments) const __attribute__((format(printf, 3, 0))) = 0;
 
     /// Defines optional instance identification for content adjustment.
     virtual const char*  traceId() const = 0;
 
     /**
-     * LOG_METHOD_TEMPLATE defines a non-virtual method using its first parameter. The method
+     * DECLARE_LOG_METHOD_TEMPLATE defines a non-virtual method using its first parameter. The method
      * defines a shared instance of a LogMsgCategory using the remaining parameters. If the
      * requested message priority is neither zero nor greater than the message priority limit, the
      * message will be logged.
      */
-#define LOG_METHOD_TEMPLATE(function, audience, classification) \
-    void function(LogMsgDetail priorityRequested, const char* format, ...) const __attribute__((format(printf, 3, 4))) \
+#define DECLARE_LOG_METHOD_TEMPLATE(function, audience, classification) \
+    void function(LogMsgDetail priorityRequested, const char* format, ...) const __attribute__((format(printf, 3, 4)));
+
+    DECLARE_LOG_METHOD_TEMPLATE(uerrlog, MSGAUD_user, MSGCLS_error);
+    DECLARE_LOG_METHOD_TEMPLATE(uwarnlog, MSGAUD_user, MSGCLS_warning);
+    DECLARE_LOG_METHOD_TEMPLATE(uproglog, MSGAUD_user, MSGCLS_progress);
+    DECLARE_LOG_METHOD_TEMPLATE(uinfolog, MSGAUD_user, MSGCLS_information);
+    DECLARE_LOG_METHOD_TEMPLATE(ierrlog, MSGAUD_programmer, MSGCLS_error);
+    DECLARE_LOG_METHOD_TEMPLATE(iwarnlog, MSGAUD_programmer, MSGCLS_warning);
+    DECLARE_LOG_METHOD_TEMPLATE(iproglog, MSGAUD_programmer, MSGCLS_progress);
+    DECLARE_LOG_METHOD_TEMPLATE(iinfolog, MSGAUD_programmer, MSGCLS_information);
+    DECLARE_LOG_METHOD_TEMPLATE(oerrlog, MSGAUD_operator, MSGCLS_error);
+    DECLARE_LOG_METHOD_TEMPLATE(owarnlog, MSGAUD_operator, MSGCLS_warning);
+    DECLARE_LOG_METHOD_TEMPLATE(oproglog, MSGAUD_operator, MSGCLS_progress);
+    DECLARE_LOG_METHOD_TEMPLATE(oinfolog, MSGAUD_operator, MSGCLS_information);
+
+#undef DECLARE_LOG_METHOD_TEMPLATE
+};
+
+#define DEFINE_LOG_METHOD_TEMPLATE(function, audience, classification) \
+    inline void ITraceLoggingComponent::function(LogMsgDetail priorityRequested, const char* format, ...) const \
     { \
         static const LogMsgCategory LMC##function(audience, classification, 1); \
```

Comment:
statics within inline functions can cause dynamically loaded shared objects to be never unloaded.  Best avoided - in this case if it was a constexpr it can probably move outside the function definition.